src/rust/engine/src/externs/mod.rs

```diff
@@ -35,15 +35,38 @@ mod stdio;
 pub mod testutil;
 pub mod workunits;
 
-pub fn register(_py: Python, m: &PyModule) -> PyResult<()> {
+pub fn register(py: Python, m: &PyModule) -> PyResult<()> {
   m.add_class::<PyFailure>()?;
+  m.add("EngineError", py.get_type::<EngineError>())?;
+  m.add("IntrinsicError", py.get_type::<IntrinsicError>())?;
+  m.add(
+    "IncorrectProductError",
+    py.get_type::<IncorrectProductError>(),
+  )?;
+
   Ok(())
 }
 
+create_exception!(native_engine, EngineError, PyException);
+create_exception!(native_engine, IntrinsicError, EngineError);
+create_exception!(native_engine, IncorrectProductError, EngineError);
+
 #[derive(Clone)]
 #[pyclass]
 pub struct PyFailure(pub Failure);
 
+#[pymethods]
+impl PyFailure {
+  fn get_error(&self, py: Python) -> PyObject {
```

Comment:
This isn't a very specific method name, given that it returns `PyObject`. It seems like in the `Failure::Throw` case, this might be any type of exception, rather than only `Error`...?

Reply:
Is `get_exception` a better name? (exception/error feels interchangeable to me)

The `Failure::Throw{ val }` is a `PyAny` - but I guess ought to only ever hold a reference to a `PyErr` (i.e. `Exception`) object. Also, for the other `Failure::` cases it currently returns an `IntrinsicError`.

Reply:
Or maybe `.get_object()`, to closer match the result type..

Reply:
It's probably fine as is. If the return type could be more specific, that would be great. But failing that, it's not a big deal.